### PR TITLE
[IOTDB-3556] Maintain onlineConfigNodes by LoadManager

### PR DESCRIPTION
**Pull request by RYH61**
1、In NodeManager：change getOnlineConfigNodes() to getRegisteredDataNodes
2、move getOnlineConfigNodes() to LoadManager

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessor.java [hunk 1]
```diff
@@ -397,7 +397,7 @@ public TSStatus removeConfigNode(TConfigNodeLocation configNodeLocation) throws
    */
   @Override
   public TSStatus stopConfigNode(TConfigNodeLocation configNodeLocation) throws TException {
-    if (!configManager.getNodeManager().getOnlineConfigNodes().contains(configNodeLocation)) {
+    if (!configManager.getLoadManager().getOnlineConfigNodes().contains(configNodeLocation)) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/ConfigNodeRPCServiceProcessor.java:400]: getRegisteredConfigNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java [hunk 2]
```diff
@@ -142,7 +142,7 @@ public boolean verifySucceed(TSStatus... status) {
    */
   public void addConsensusGroup(TConfigNodeLocation tConfigNodeLocation) {
     List<TConfigNodeLocation> configNodeLocations = new ArrayList<>();
-    configNodeLocations.addAll(configManager.getNodeManager().getOnlineConfigNodes());
+    configNodeLocations.addAll(configManager.getLoadManager().getOnlineConfigNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:145]: getRegisteredConfigNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java [hunk 3]
```diff
@@ -339,18 +339,18 @@ private void storeConfigNode() throws IOException {
...
       systemProperties.store(fileOutputStream, "");
     }
   }
 
-  public List<TConfigNodeLocation> getOnlineConfigNodes() {
+  public List<TConfigNodeLocation> getRegisteredDataNodes() {
     List<TConfigNodeLocation> result;
     configNodeInfoReadWriteLock.readLock().lock();
     try {
       // TODO: Check ConfigNode status, ensure the returned ConfigNode isn't removed
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:352]: This typo can be removed now

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 4]
```diff
@@ -91,13 +92,17 @@ public class LoadManager {
   private Future<?> currentHeartbeatFuture;
   private int balanceCount = 0;
 
+  /** Online ConfigNodes * */
+  private List<TConfigNodeLocation> onlineConfigNodes;
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:96]: We shouldn't add another field in LoadManager

Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:96]: We shouldn't add another field in LoadManager

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 5]
```diff
@@ -206,7 +211,7 @@ private void heartbeatLoopBody() {
       // Send heartbeat requests to all the online DataNodes
       pingOnlineDataNodes(getNodeManager().getOnlineDataNodes(-1));
       // Send heartbeat requests to all the online ConfigNodes
-      pingOnlineConfigNodes(getNodeManager().getOnlineConfigNodes());
+      pingOnlineConfigNodes(getNodeManager().getRegisteredDataNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:214]: This interface should be renamed to pingRegisteredConfigNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 6]
```diff
@@ -281,6 +286,19 @@ public void removeNodeHeartbeatHandCache(Integer nodeId) {
...
+    getNodeManager()
+        .getRegisteredDataNodes()
+        .forEach(
+            registerConfigNode -> {
+              if (heartbeatCacheMap.get(registerConfigNode.getConfigNodeId()).getNodeStatus()
+                  == NodeStatus.Unknown) {
+                onlineConfigNodes.remove(registerConfigNode);
+              }
+            });
+    return onlineConfigNodes;
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:299]: New List<TConfigNodeLocation> to store the result, thus we don't have to maintain the onlineConfigNodes field in LoadManager

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 7]
```diff
@@ -117,7 +118,7 @@ public DataSet registerDataNode(RegisterDataNodePlan req) {
     }
 
     dataSet.setDataNodeId(req.getInfo().getLocation().getDataNodeId());
-    dataSet.setConfigNodeList(nodeInfo.getOnlineConfigNodes());
+    dataSet.setConfigNodeList(getLoadManager().getOnlineConfigNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:121]: getRegisteredConfigNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 8]
```diff
@@ -177,7 +178,7 @@ public TConfigNodeRegisterResp registerConfigNode(TConfigNodeRegisterReq req) {
     resp.setPartitionRegionId(
         getConsensusManager().getConsensusGroupId().convertToTConsensusGroupId());
 
-    resp.setConfigNodeList(nodeInfo.getOnlineConfigNodes());
+    resp.setConfigNodeList(getLoadManager().getOnlineConfigNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:181]: getRegisteredConfigNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 9]
```diff
@@ -200,14 +201,16 @@ public TSStatus removeConfigNode(RemoveConfigNodePlan removeConfigNodePlan) {
...
           return new TSStatus(TSStatusCode.REMOVE_CONFIGNODE_FAILED.getStatusCode())
               .setMessage(
                   "Remove ConfigNode failed because there is only one ConfigNode in current Cluster.");
         }
 
         // Check whether the onlineConfigNodes contain the ConfigNode to be removed.
-        if (!getOnlineConfigNodes().contains(removeConfigNodePlan.getConfigNodeLocation())) {
+        if (!getLoadManager()
+            .getOnlineConfigNodes()
+            .contains(removeConfigNodePlan.getConfigNodeLocation())) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:213]: getRegisteredConfigNodes and modify the typo here

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConsensusManager.java [hunk 10]
```diff
@@ -168,7 +169,7 @@ public TConfigNodeLocation getLeader() {
     for (int retry = 0; retry < 50; retry++) {
       Peer leaderPeer = consensusImpl.getLeader(consensusGroupId);
       if (leaderPeer != null) {
-        List<TConfigNodeLocation> onlineConfigNodes = getNodeManager().getOnlineConfigNodes();
+        List<TConfigNodeLocation> onlineConfigNodes = getLoadManager().getOnlineConfigNodes();
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConsensusManager.java:172]: getRegisteredConfigNode

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 11]
```diff
@@ -174,7 +174,7 @@ public DataSet registerDataNode(RegisterDataNodePlan registerDataNodePlan) {
     } else {
       DataNodeConfigurationResp dataSet = new DataNodeConfigurationResp();
       dataSet.setStatus(status);
-      dataSet.setConfigNodeList(nodeManager.getOnlineConfigNodes());
+      dataSet.setConfigNodeList(loadManager.getOnlineConfigNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:177]: getRegisteredConfigNodes

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL~

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java (hunk 2 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 12]
```diff
@@ -200,14 +201,14 @@ public TSStatus removeConfigNode(RemoveConfigNodePlan removeConfigNodePlan) {
     if (removeConfigNodeLock.tryLock()) {
       try {
         // Check ConfigNodes number
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:203]: Fix typo

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 13]
```diff
@@ -200,14 +201,14 @@ public TSStatus removeConfigNode(RemoveConfigNodePlan removeConfigNodePlan) {
...
-        if (getOnlineConfigNodes().size() <= 1) {
+        if (getLoadManager().getOnlineConfigNodes().size() <= 1) {
           return new TSStatus(TSStatusCode.REMOVE_CONFIGNODE_FAILED.getStatusCode())
               .setMessage(
                   "Remove ConfigNode failed because there is only one ConfigNode in current Cluster.");
         }
 
         // Check whether the onlineConfigNodes contain the ConfigNode to be removed.
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:210]: Fix typo

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 14]
```diff
@@ -232,7 +234,7 @@ private void heartbeatLoopBody() {
       // Send heartbeat requests to all the online DataNodes
       pingOnlineDataNodes(getNodeManager().getOnlineDataNodes(-1));
       // Send heartbeat requests to all the online ConfigNodes
-      pingOnlineConfigNodes(getNodeManager().getOnlineConfigNodes());
+      pingOnlineConfigNodes(getNodeManager().getRegisteredConfigNodes());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:237]: pingRegisteredConfigNodes

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

**Review by CRZbulabula — APPROVED** [review]
LGTM

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 15]
```diff
@@ -232,7 +234,7 @@ private void heartbeatLoopBody() {
       // Send heartbeat requests to all the online DataNodes
       pingOnlineDataNodes(getNodeManager().getOnlineDataNodes(-1));
       // Send heartbeat requests to all the online ConfigNodes
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:236]:
```suggestion
      // Send heartbeat requests to all the registered ConfigNodes
```

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 16]
```diff
@@ -287,7 +289,7 @@ private void pingOnlineDataNodes(List<TDataNodeInfo> onlineDataNodes) {
    *
    * @param onlineConfigNodes ConfigNodes that currently online
    */
-  private void pingOnlineConfigNodes(List<TConfigNodeLocation> onlineConfigNodes) {
+  private void pingRegisteredConfigNodes(List<TConfigNodeLocation> onlineConfigNodes) {
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:292]:
```suggestion
  private void pingRegisteredConfigNodes(List<TConfigNodeLocation> registeredConfigNodes) {
```

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 17]
```diff
@@ -287,7 +289,7 @@ private void pingOnlineDataNodes(List<TDataNodeInfo> onlineDataNodes) {
    *
    * @param onlineConfigNodes ConfigNodes that currently online
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:290]:
```suggestion
   * @param registeredConfigNodes ConfigNodes that registered in cluster
```

Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:290]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java (hunk 16 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:292]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java (hunk 15 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:236]: fixed~

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**